### ISSUE TITLE
Developers overriding production location for retrieving conditions in configurations

### ISSUE THREAD

**Issue by davidlange6** [cms-sw/cmssw#27393]
Breaks centrally determined location for retrieving conditions (which of course can change even if it rarely does). All of these should rely on the global tag to determine from where the conditions are to be taken (or they should not be loaded into production workflows)

https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/L1Trigger/L1TMuon/python/simGmtStage2Digis_cfi.py#L23

https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L41

https://github.com/cms-sw/cmssw/blob/8587c76d0a2ee041152b0c0ac289d851a13ef247/L1Trigger/L1TCalorimeter/python/simDigis_cff.py#L49

**Comment by davidlange6** [comment]
assign l1,db,reco

**Comment by davidlange6** [comment]
assign reconstruction

**Comment by ggovi** [comment]
My personal opinion is that a cfg file containing a Condition Customization ( defined with one or more toGet statements ) should be NEVER invoked ( directly or included ) in a production workflow, since:
- it overrides the mapping record/tag provided by the official Global Tag, referencing some Tag that could stay outside the control ( in terms of testing and validation ) 
- it potentially breaks reproducibility
For this reason, I think that the first measure required is to find out the reason of the concerned customisation, and remove them from the production workflows.
This will potentially imply:
- to replace some tag in existing GTs (?)
- to create additional GTs (?)
I think AlCa people should be involved in this review procedure

**Comment by slava77** [comment]
for cmssw/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py
@mbluj @steggema @swozniewski please comment and perhaps follow up with a resolution

**Comment by fabiocos** [comment]
for the L1Trigger instances: @rekovic could you please address this issue?

**Comment by ggovi** [comment]
Can we add alca in this loop?

**Comment by kpedro88** [comment]
assign alca

**Comment by mbluj** [comment]
Hello,
comments on https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L41
1. TauID payloads, i.e. BDT training files and pt-dependent WP definitions, are mapped via python configuration file `RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py` rather than via a GlobalTag. In fact the payloads do not define what one could call "conditions", but define a version of a give tau identification algorithm.
2. This particular tool (`runTauIdMVA.py`) is used to add new tauIDs to `process` object and then potentially to a given workflow - PAT/MiniAOD or post-PAT/MiniAOD. Until 10_6_X the tool has been used only by final users to produce their analysis ntuples on top of MiniAOD. Since 10_6_X the tool is used to add one particular tauID to MiniAOD (namely DeepTauID). 
3. As it is handy to maintain definition of tauIDs in one place it is considered to use the tool also to define upgraded tauIDs for nanoAOD.
4. Concerning retrieving conditions in #L41 of the tool - it is inactive. It can be activated only if an nonempty connection address is defined in #L31. The reason of such implementation is to make test of payloads from preparational DB simpler.

**Comment by slava77** [comment]
+1

the only case affecting reco is clarified in https://github.com/cms-sw/cmssw/issues/27393#issuecomment-506741888
this looks like a reasonable use case for private tests, which does not affect the production setup.

**Comment by davidlange6** [comment]
these are the payloads that were needed to run digi+reco without frontier - it does contradict the conclusion of @slava77 and @mbluj 

https://gitlab.cern.ch/hep-benchmarks/hep-workloads/blob/cae5bd13ac1aaf3df01f224db642f2c00df26d93/cms/reco/cms-reco/generate_GlobalTag.sh

**Comment by slava77** [comment]
> generate_GlobalTag.sh

please elaborate what's in this file.
Thank you.

**Comment by davidlange6** [comment]
all the payloads not in the GT that without which a digi-reco-miniaod workflow will not run

**Comment by davidlange6** [comment]
which I think is not to say (if or if not) they are actually read and used..but rather that our applications depend on them being present to function - perhaps due to purely technical reasons

**Comment by davidlange6** [comment]
(sorry, didn't mean to close this)

**Comment by slava77** [comment]
I think that the origin is in https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py#L5

not in the https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L41

**Comment by mbluj** [comment]
> which I think is not to say (if or if not) they are actually read and used..but rather that our applications depend on them being present to function - perhaps due to purely technical reasons

Sorry, but I am lost... The "tau" payloads are used by tauIDs, but they are not mapped via a GlobalTag, but as defined in a python configuration file (cfi). In my understanding the original issue was the fact that there are in CMSSW configuration files or python tools modifying location for retrieving such payloads, but as I already wrote it is **not the case** for the "tau" payloads.

**Comment by mbluj** [comment]
> I think that the origin is in
> 
> https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py#L5
> 
> not in the
> 
> https://github.com/cms-sw/cmssw/blob/82c4a527471bca6fc4897fd4feb3f7d7fcb73928/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L41

Yes, I confirm this. If one finds it potentially danger that the connection string of the "tauDB service" (ESSource) is defined independently from the service (ESSource) used by GlobalTag we can correct it by cloning the service used by GlobalTag, so that the tauDB ESSource will be linked to GlobalTag's one.

**Comment by davidlange6** [comment]
so the concept of reading db payloads outside of the GT was discussed and approved by alca/db or this was just done without consulting them?

**Comment by mbluj** [comment]
> so the concept of reading db payloads outside of the GT was discussed and approved by alca/db or this was just done without consulting them?

It is in CMSSW since 2013 or so... I guess it was agreed, but as I have not participated directly in this development I cannot be entirely sure. Regardless, nowadays the question can be if it is OK for alca/db to stay with what is in CMSSW for long time, and if it is not fine how to migrate this to GT (with some schedule).

**Comment by mbluj** [comment]
BTW, payloads with training weights of BDT-based tauIDs were put to DB to replace the (.root) files containing the weights as the files are not easy for handling. Currently training files for DNN-based algorithms (b-tagging, tauID, etc.) are again outside DB and hence not controlled by GT...

**Comment by slava77** [comment]
https://hypernews.cern.ch/HyperNews/CMS/get/swDevelopment/3215/1/2/2/1/1/2/1/1.html

**Comment by slava77** [comment]
@mmusich do you have a record of what happened since after https://hypernews.cern.ch/HyperNews/CMS/get/swDevelopment/3215/1/2/2/1/1/2/1/1.html
?
Please check/comment.
Thank you.

**Comment by mbluj** [comment]
> https://hypernews.cern.ch/HyperNews/CMS/get/swDevelopment/3215/1/2/2/1/1/2/1/1.html

Thank you Slava! Now the situation is more or less clear: the tauID payloads are in appropriate DB, but still not under GT control. Can mapping of the payloads added to GT or are already there? When it is done we can remove mapping from cfi.

**Comment by mmusich** [comment]
Hi @slava77, so, as far as I recall the problem with this issue is that there is a very large amount of tau (payloads, labels) pairs to manage in GT.  Including them in GT would roughly speaking double the size of the Global Tag. 
At the time of the HN message with the old condition management system including 150 tag/labels would have taken simply too much time (not to speak of the fatigue of making sure the right couplings are done). Now, with the new CondDB browser it can be done quickly, but I would still have a concern about the high chances of making mistakes when doing the queueing. 
As far as I remember these conditions are changed fairly frequently and it would be a good investment for the future if the Tau POG tries to re-haul they CondFormat data structures in order to limit the number of labels used (that's what we ended up suggesting in 2015).  
@mbluj 
 > Can mapping of the payloads added to GT or are already there?

the mapping can be done in the GT, but someone from tau POG needs to do all the queuing for all 150 payloads for all the relevant production queues  (trust me, it won't be a fun afternoon ;))

**Comment by mbluj** [comment]
> Hi @slava77, so, as far as I recall the problem with this issue is that there is a very large amount of tau (payloads, labels) pairs to manage in GT. Including them in GT would roughly speaking double the size of the Global Tag.
> At the time of the HN message with the old condition management system including 150 tag/labels would have taken simply too much time (not to speak of the fatigue of making sure the right couplings are done). Now, with the new CondDB browser it can be done quickly, but I would still have a concern about the high chances of making mistakes when doing the queueing.
> As far as I remember these conditions are changed fairly frequently and it would be a good investment for the future if the Tau POG tries to re-haul they CondFormat data structures in order to limit the number of labels used (that's what we ended up suggesting in 2015).
> @mbluj
> 
> > Can mapping of the payloads added to GT or are already there?
> 
> the mapping can be done in the GT, but someone from tau POG needs to do all the queuing for all 150 payloads for all the relevant production queues (trust me, it won't be a fun afternoon ;))

I am dealing with the cfi file, so I can easily agree that the work is not fun, but when it is done it will be much less work in future. Anyway, LS2 is good occasion to clean tauID payloads - many are kept only for backward compatibility and there is not need to add them to GT -- to be discussed with tau conveners @steggema, @rmanzoni, @roger-wolf. Maybe CondFormat can be redone too, but it surely requires more time.

**Comment by mbluj** [comment]
Just to let you know: tau POG is starting an effort to finally map our payloads via GT. This will go together with review of tauIDs (this should reduce no. of payloads to map) which is anyway scheduled for this Summer. I hope you will hear more about this at next Reco/AT meeting.

**Comment by fabiocos** [comment]
@rekovic we still need to clarify the access of conditions outside GT in L1 configurations, could you please comment?

**Comment by rekovic** [comment]
@fabiocos
At least in the case of https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/L1Trigger/L1TMuon/python/simGmtStage2Digis_cfi.py#L23
these are "Prototype" parameters which are parameters not transferred to the DB in the process of O2O.
The emulator in turn needs both the the Prototype and the  O2O-ed parameters.
@thomreis 

In the case of 
https://github.com/cms-sw/cmssw/blob/8587c76d0a2ee041152b0c0ac289d851a13ef247/L1Trigger/L1TCalorimeter/python/simDigis_cff.py#L49
we are investigating, currently considering removing under the hypothesis that the needed payload is in the GT already.   Will need to follow up.
@bundocka

**Comment by ggovi** [comment]
Within the AlCaDB team we are currently discussing this issue. In parallel to the review and clean up, which is being addressed here, we believe it would be beneficial to introduce some policy to better regulate the use of customisations, especially for what concerns the production workflows like HLT, Express, Prompt Reco.  Two aspects can be considered:
1. How to ensure the control of *allowed* customisations, if any.  This could be achieved by force the customisation to be declared in files under the github AlCa signature. We still need to identify the best way to implement this.
2. How to protect the production workflows from *undesired* customisations, that could appear in nested cff/cfg files outside the AlCa signature. On this scope, we could introduce a special flag, something like “is_production” ( we are wondering is something similar already exists… ), that automatically - when set - could reject the customisations during the loading of the Global Tag. The rejection could consist simply in ignoring them, or additionally in throwing an error, by choice.   
We are currently only at level of brainstorming, so please provide feedback ( ideas are also welcome )  for these proposals.

**Comment by christopheralanwest** [comment]
> these are the payloads that were needed to run digi+reco without frontier - it does contradict the conclusion of @slava77 and @mbluj
> 
> https://gitlab.cern.ch/hep-benchmarks/hep-workloads/blob/cae5bd13ac1aaf3df01f224db642f2c00df26d93/cms/reco/cms-reco/generate_GlobalTag.sh

@davidlange6 Could you clarify if the ability to run digi+reco without frontier is expected to be supported absolutely? This isn't a normal use case. If a compute node does not have the connectivity required to communicate with frontier via a squid, how would data get in or out of the site? Is the requirement to be able to run in this way needed to support open data efforts?

In the case of HLT, Express and Prompt reconstruction, using an exported global tag can never be supported because the corresponding global tags are not fixed. For these global tags, the snapshot time is set to infinity to allow new payloads can be appended to tags.

While it seems clear that a mechanism to prevent overrides of condition data in production workflows is desirable, it's less obvious to me that such overrides need to be forbidden for Phase-2 workflows. AlCa currently does not review fake conditions such as those used by the Phase-2 inner tracker or Phase-2 HCAL as they can be modified in python configurations in packages that do not require AlCa signature. It is somewhat arbitrary to require that AlCa review fake conditions the instant that they are implemented as tags rather than as python configuration. My personal opinion is that AlCa should largely defer to reconstruction, upgrade, simulation and subdetector experts for review of changes to Phase-2 scenarios. We don't have the expertise needed to fully review such fake conditions.

To review specific cases mentioned above:

- tau payloads: in a separate thread, tau experts have shared plans to incorporate tau payloads into 11_0_X global tags. We will aid with this process once 11_0_X queues have been created
- L1T payloads: I don't understand why the need to have these payloads outside a global tag. The list of global tags in which the `L1TCaloParams_static_CMSSW_9_2_10_2017_v1_8_2_updateHFSF_v6MET` tag is included [seems particularly haphazard](https://cms-conddb.cern.ch/cmsDbBrowser/search/Prod/L1TCaloParams_static_CMSSW_9_2_10_2017_v1_8_2_updateHFSF_v6MET); currently only 2016 MC scenarios use this global tag whereas up to 94X, all MC scenarios except "2021" used this tag. Could @rekovic @thomreis and @bundocka  follow up at a later AlCa meeting?
- Phase-II tracker: I don't see the need for AlCa to maintain five different global tag queues that differ only by Phase-II tracker geometry as in #27517. I would prefer their #27516, even though it conflicts with the goal of this issue.

For my education: why does the connect string not have a sensible default, such as some value that is aliased to the actual FrontierProd connect string? Is this done to avoid implicit behavior or are there other motivations as well?

**Comment by mmusich** [comment]
@christopheralanwest 

> Phase-II tracker: I don't see the need for AlCa to maintain five different global tag queues that differ only by Phase-II tracker geometry as in #27517. I would prefer their #27516, even though it conflicts with the goal of this issue.

just as a clarification, also #27517 is based on the same append-payload mechanism as #27516, only it does it in a different way not passing from an additional `ESSource` (it uses the one of the Global Tag).
In any case AlCa would not be asked to maintain five different global tags in either of the two options, we can just agree that we (Tracker) manage that part of the conditions and update the tag names accordingly when needed.

**Comment by davidlange6** [comment]
hi @christopheralanwest 

I think you have several topics

0) is sqlite a good solution for data taking operations? no.. for the reasons you state. (and others)

1) should alcadb continue to support sqlite access to conditions? There are use cases like opendata, testing applications in self contained environment and developers testing conditions updates. 

2) should a global tag contain all of the database objects needed by cms to run its production applications? This was the topic of this issue. I think "yes". But if there is a policy change it would be useful to know the new policy. Indeed, this does require the release manager (and others) to flag when code is coming in that overrides this policy. 

3) as for phase II - there are various solutions that don't require separate GTs (using labels, iovs, etc)... Alca should think of how it wants to support needed use cases.

**Comment by kpedro88** [comment]
I think using labels is the only way we'll ever be able to build up a complete Phase 2 GT before installation actually occurs.

**Comment by davidlange6** [comment]
i can think of at least one other... but anyway, its >0 options that don't require multiple gts....

**Comment by christopheralanwest** [comment]
First of all, tau experts have queued their tags to 11_0_X queues so that these tags can be included in global tags. Tests to ensure that obtaining the tags from the global tag gives equivalent results to the current method are underway.

One issue that hasn't been discussed here is maintainability by AlCa. AlCa currently actively maintains 35 different global tag queues and most of the tasks for global tag maintenance are performed via a web interface. Since there are multiple steps in creating a global tag, some of which involve non-negligible time waiting for the server to process and respond, even the most trivial change to all queues already takes a minimum of 1-2 hours of highly repetitive work. With so many queues, it is easy for a mistake to creep in somewhere when a nontrivial change needs to be made.

To give an example, some of the GTs that AlCa maintain are intended to always maintain a constant tag difference with respect to another queue, but the global tag queues need to be maintained separately:

https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_dataRun2_relval_v16/106X_dataRun2_v17
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_upgrade2018_realistic_HEfail_v6/106X_upgrade2018_realistic_v6
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_upgrade2018cosmics_realistic_deco_v3/106X_upgrade2018cosmics_realistic_peak_v3
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_mc2017cosmics_realistic_deco_v2/106X_mc2017cosmics_realistic_peak_v2
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_upgrade2018cosmics_realistic_deco_v3/106X_upgrade2018cosmics_realistic_peak_v3
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_dataRun2_PromptLike_HEfail_v9/106X_dataRun2_PromptLike_v10

All of these should be maintained with symbolic GTs to ensure that a constant tag difference is maintained. In fact, the last diff most likely corresponds to an instance in which the global tags were maintained incorrectly; I don't see any reason that there should be a difference between these two tags in non-HCAL tags. The suggestion to introduce at least four new cases in which AlCa should maintain new GTs with constant tag difference is a step in the wrong direction. We should not be relying on humans to repeatedly do the same task when it can trivially be handled once and for all programatically.

In the long term, support for symbolic GTs as already implemented in https://github.com/cms-sw/cmssw/blob/master/Configuration/HLT/python/autoCondHLT.py and as proposed in PR #27517 seems necessary for maintainability. PdmV experts tell me that symbolic tags are supported, though I don't know that they have ever been used in practice (I also don't know how the colon in "auto:" is handled).

It's true that in the case of the inner tracker conditions, it may be technically possible to propagate information about the geometry into every single location in the code where conditions are requested so that labelled conditions can be used instead. But this would be an ugly hack; a solution based on conditions is preferable.

Perhaps better solutions for AlCa maintainability are possible in the long term. Kevin and I discussed the possibility to have a dedicated AlCa/Upgrade L3 who will focus on upgrade-specific conditions. Another possibility (not at all related to the thread here) would be the introduction year-dependent MC (run-dependent MC with one IOV per year) for legacy conditions.

**Comment by fabiocos** [comment]
referring to the comment https://github.com/cms-sw/cmssw/issues/27393#issuecomment-510075816 by @ggovi:

- #27517 goes in the direction of suggestion 1), i.e. the proposed customization for phase2 relvals falls within ```Configuration/AlCa``` under the full control of the AlCaDB coordinators;

- concerning the point 2) I believe that this proposed flag should be active within ConfigBuilder.py, which is the engine that builds our regular workflows (both for  production and relval). At present I do not see anything of this kind, but the way I see it we would need a flag to veto conditions customizations (through any ```toGet``` method). This needs to be implemented.

**Comment by christopheralanwest** [comment]
In a private email thread, @bundocka suggested that the L1T condition overrides listed in https://github.com/cms-sw/cmssw/issues/27393#issue-461958525 were no longer needed. @rekovic Could you please confirm and submit a PR to remove the overrides?

**Comment by tvami** [comment]
Hi @rekovic @bundocka my understanding is that this issue is waiting on submitting a PR to remove the overrides. Do you have an ETA for that? Thanks!

**Comment by tvami** [comment]
Hi All, just FYI we will be discussing this at the meeting https://indico.cern.ch/event/1069945/#11-removal-of-unused-tau-tags

**Comment by tvami** [comment]
> Hi @rekovic @bundocka my understanding is that this issue is waiting on submitting a PR to remove the overrides. Do you have an ETA for that? Thanks!

Hi @rekovic @bundocka , could we maybe schedule a discussion about the L1T tags at the same day as the tau tags, Sept 13?

**Comment by bundocka** [comment]
Hi, @rekovic please can you follow up on this.

**Comment by tvami** [comment]
Hi @lathomas could you please look into the L1T side of this issue? Thanks!

**Comment by bundocka** [comment]
Laurent is away at the moment. @rekovic needs to look at this. I will contact him privately.

**Comment by tvami** [comment]
Thanks @bundocka !

**Comment by rekovic** [comment]
Looking again at the issue.
There should be no obvious reason why these conditions are being fetched as opposed to being part of a GT(s).
https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/L1Trigger/L1TMuon/python/simGmtStage2Digis_cfi.py#L23
https://github.com/cms-sw/cmssw/blob/8587c76d0a2ee041152b0c0ac289d851a13ef247/L1Trigger/L1TCalorimeter/python/simDigis_cff.py#L49
Checking again with GlobalMuon and Calo experts and to be followed up with a PR.  Will keep posted.

**Comment by tvami** [comment]
> For the L1T part please see #35554

Actually that never converged so see https://github.com/cms-sw/cmssw/pull/36408 instead

**Comment by tvami** [comment]
+db
 * just for the record

**Comment by davidlange6** [comment]
reopening as only 1/2 done..

**Comment by mmusich** [comment]
> reopening as only 1/2 done..

what is missing?

**Comment by davidlange6** [comment]
the tau issue above has been rediscovered today.. I don't think it was ever fixed (happy to be wrong)

**Comment by vlimant** [comment]
assign xpog

**Comment by mmusich** [comment]
> the tau issue above has been rediscovered today.. I don't think it was ever fixed (happy to be wrong)

I think it only affects relatively old release (pre Run3). Should the changes be backported to (at least) the UL releases?

**Comment by mmusich** [comment]
@tvami @mbluj 

Was the proposal at https://indico.cern.ch/event/1069945/#11-removal-of-unused-tau-tags implemented?

**Comment by mbluj** [comment]
> @tvami @mbluj
> 
> Was the proposal at https://indico.cern.ch/event/1069945/#11-removal-of-unused-tau-tags implemented?

Yes, I believe it was implemented, but let me find appropriate PR. Anyway, I think it is still possible to rerun old tauIDs within NanoAOD sequeces with appropriately old eras. 

Regardless, which release is affected by this issue. What was actual `cmsDriver` command used to produce configuration. It could be that some unexpected set of parameters was used.

**Comment by vlimant** [comment]
wf number `2500.001` in the nano matrix, at least

**Comment by katilp** [comment]
The cmsDriver command is [here](https://github.com/cms-opendata-analyses/PFNanoProducerTool/blob/master/PFNanoDrivers.sh)

```
# Data 2016
cmsDriver.py pfnano_data_2016UL_OpenData --data --eventcontent NANOAODSIM --datatier NANOAODSIM --step NANO \
--conditions 106X_dataRun2_v37   --era Run2_2016,run2_nanoAOD_106Xv2 \
--customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)))" --nThreads 4 \
-n 100 --filein /store/data/Run2016H/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/676E37D2-044C-D346-92D9-A127A55FD279.root --fileout file:nano_data2016.root \
--customise PhysicsTools/PFNano/pfnano_cff.PFnano_customizeData_onlyaddPFcands$NOINPUTS  $NO_EXEC $PARALLEL
```

**Comment by mbluj** [comment]
As far as I can find the cleaning was performed in https://github.com/cms-sw/cmssw/pull/37091 (12_3). Further cleaning will require removal of options for old eras and breaking backward compatibility with old miniAOD. As the backward is compatibility is anyhow just theoretical this is desirable, I think. On the other hand, I am not in favor to backport this down to UL release series.

**Comment by mbluj** [comment]
> The cmsDriver command is [here](https://github.com/cms-opendata-analyses/PFNanoProducerTool/blob/master/PFNanoDrivers.sh)
> 
> ```
> # Data 2016
> cmsDriver.py pfnano_data_2016UL_OpenData --data --eventcontent NANOAODSIM --datatier NANOAODSIM --step NANO \
> --conditions 106X_dataRun2_v37   --era Run2_2016,run2_nanoAOD_106Xv2 \
> --customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)))" --nThreads 4 \
> -n 100 --filein /store/data/Run2016H/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/676E37D2-044C-D346-92D9-A127A55FD279.root --fileout file:nano_data2016.root \
> --customise PhysicsTools/PFNano/pfnano_cff.PFnano_customizeData_onlyaddPFcands$NOINPUTS  $NO_EXEC $PARALLEL
> ```

Which CMSSW release is used?

**Comment by katilp** [comment]
> > The cmsDriver command is [here](https://github.com/cms-opendata-analyses/PFNanoProducerTool/blob/master/PFNanoDrivers.sh)
> > ```
> > # Data 2016
> > cmsDriver.py pfnano_data_2016UL_OpenData --data --eventcontent NANOAODSIM --datatier NANOAODSIM --step NANO \
> > --conditions 106X_dataRun2_v37   --era Run2_2016,run2_nanoAOD_106Xv2 \
> > --customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)))" --nThreads 4 \
> > -n 100 --filein /store/data/Run2016H/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/676E37D2-044C-D346-92D9-A127A55FD279.root --fileout file:nano_data2016.root \
> > --customise PhysicsTools/PFNano/pfnano_cff.PFnano_customizeData_onlyaddPFcands$NOINPUTS  $NO_EXEC $PARALLEL
> > ```
> 
> Which CMSSW release is used?

CMSSW_10_6_30

**Comment by mmusich** [comment]
In any case the specific tag (`RecoTauTag_againstMuonMVAv1`) that is causing the exception reported at https://cms-talk.web.cern.ch/t/condition-database-access-outside-of-gt-for-nano-production/33715 is not in any GlobalTag:

```
$ conddb listGTsForTag RecoTauTag_againstMuonMVAv1
[2024-01-26 16:53:58,755] INFO: Connecting to pro [frontier://PromptProd/cms_conditions]
GT_name  Tag_name  record  label   
-------  --------  ------  -----   
```

so it would need to be included (setting aside changing the code supplying it).

**Comment by mbluj** [comment]
I would better remove usage of this (and possibly other) depreciated tauIDs from 10_6 for open-data production if possible. This particular tauID (`againstMuonMVAv1`) was not supported and I think not stored in official NanoAOD at all.
How urgent it is? I have other items on the plate, but I can change priorites if needed. 

Regardless, I think that there are at least two technical problems to solve:
* remove depreciated tauID from NanoAOD in master, by removing support for eras older than Run-2 UL `run2_nanoAOD_106Xv2`,
* fix the issue in UL release series (10_6), either by modifications of tau nano configuration or adding problematic payload to GT (or both). If fixes in cmssw are possible (requires at the end publication of a new release) I will favor it even if it requires my work... Fix for 10_6 are not necessary a (simple) backport of fix for master.

**Comment by mmusich** [comment]
> This particular tauID (againstMuonMVAv1) was not supported and I think not stored in official NanoAOD at all.

it begs the question why such unsupported conditions are then required for the open-data.
Is it coming from `--customise PhysicsTools/PFNano/pfnano_cff.PFnano_customizeData_onlyaddPFcands` ?

**Comment by mbluj** [comment]
> > This particular tauID (againstMuonMVAv1) was not supported and I think not stored in official NanoAOD at all.
> 
> it begs the question why such unsupported conditions are then required for the open-data. Is it coming from `run2_nanoAOD_106Xv2` ?

I do not think so, I suppose it is due to customization (not maintained by TAU).

**Comment by katilp** [comment]
If these changes are expected to affect the output we would like to have them as soon as possible for testing and eventually reprocessing of the example PFNano datasets that will go to the open data release. 
Changing the release has a big impact on the work to be done as well, but we will need to do much of that work for a new GT in any case.

**Comment by mmusich** [comment]
For my edification why is CMS releasing open data related to a custom wotkflow (requiring unsupported conditions)?

**Comment by katilp** [comment]
> For my edification why is CMS releasing open data related to a custom wotkflow (requiring unsupported conditions)?

Producing NanoAOD enriched with the PF candidates is of major interest to the open data user community. That is why we are providing that example. We do not use unsupported conditions on purpose. We are more than happy to use supported conditions, please let us know how to modify the workflow or point us to the relevant documentation. Thank you, your help will be appreciated🙂!

**Comment by mbluj** [comment]
@katilp Executing your cmsDriver command I got the following error:
```
ImportError: No module named PFNano.pfnano_cff

```
And indeed there is not `PhysicsTools/PFNano/` package in 10_6_30 and customization cff therein. Could you please provide a full recipe for copy and paste?

In parallel I checked that this configuration without the customization works without problems. This highly probably mean that the issue is caused by the customization.

I will back to it at the beginning of the next week, but I am not sure if I will have time on Monday due to other duties.

**Comment by katilp** [comment]
> > For my edification why is CMS releasing open data related to a custom wotkflow (requiring unsupported conditions)?
> 
> Producing NanoAOD enriched with the PF candidates is of major interest to the open data user community. That is why we are providing that example. We do not use unsupported conditions on purpose. We are more than happy to use supported conditions, please let us know how to modify the workflow or point us to the relevant documentation. Thank you, your help will be appreciated🙂!

Also please note that this error appears without any customization and has nothing to do with open data, other than open data preparations helping CMS to improve the codebase ❤️

**Comment by mmusich** [comment]
> Also please note that this error appears without any customization and has nothing to do with open data, other than open data preparations helping CMS to improve the codebase 

this seems in contradiction with https://github.com/cms-sw/cmssw/issues/27393#issuecomment-1912456271

**Comment by mbluj** [comment]
> > > For my edification why is CMS releasing open data related to a custom wotkflow (requiring unsupported conditions)?
> > 
> > 
> > Producing NanoAOD enriched with the PF candidates is of major interest to the open data user community. That is why we are providing that example. We do not use unsupported conditions on purpose. We are more than happy to use supported conditions, please let us know how to modify the workflow or point us to the relevant documentation. Thank you, your help will be appreciated🙂!
> 
> Also please note that this error appears without any customization and has nothing to do with open data, other than open data preparations helping CMS to improve the codebase ❤️

As written in the comment above: I was not able to reproduce the issue w/o the customization, i.e. executing the following
```
cmsrel CMSSW_10_6_30
cd CMSSW_10_6_30/src
cmsenv
cmsDriver.py data_2016UL_OpenData --data --eventcontent NANOAODSIM --datatier NANOAODSIM --step NANO --conditions 106X_dataRun2_v37   --era Run2_2016,run2_nanoAOD_106Xv2 --customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)))" --nThreads 4 -n 100 --filein /store/data/Run2016H/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/676E37D2-044C-D346-92D9-A127A55FD279.root --fileout file:nano_data2016_nopf.root  --no_exec
voms-proxy-init --voms cms
cmsRun data_2016UL_OpenData_NANO.py
```
The cmsDriver command with customization does not work for me.

**Comment by katilp** [comment]
> @katilp Executing your cmsDriver command I got the following error:
> 
> ```
> ImportError: No module named PFNano.pfnano_cff
> ```
> 
> And indeed there is not `PhysicsTools/PFNano/` package in 10_6_30 and customization cff therein. Could you please provide a full installation recipe for copy and paste to reproduce the issue?
> 
> In parallel I checked that this configuration without the customization works without problems. This highly probably mean that the issue is caused by the customization.
> 
> I will back to it at the beginning of the next week, but I am not sure if I will have time on Monday due to other duties.

Thank you!

For the sake of brevity, removing the customization:

```
import FWCore.ParameterSet.Config as cms

from Configuration.Eras.Era_Run2_2016_cff import Run2_2016
from Configuration.Eras.Modifier_run2_nanoAOD_106Xv2_cff import run2_nanoAOD_106Xv2

process = cms.Process('NANO',Run2_2016,run2_nanoAOD_106Xv2)

# import of standard configurations
process.load('Configuration.StandardSequences.Services_cff')
process.load('SimGeneral.HepPDTESSource.pythiapdt_cfi')
process.load('FWCore.MessageService.MessageLogger_cfi')
process.load('Configuration.EventContent.EventContent_cff')
process.load('Configuration.StandardSequences.GeometryRecoDB_cff')
process.load('Configuration.StandardSequences.MagneticField_AutoFromDBCurrent_cff')
process.load('PhysicsTools.NanoAOD.nano_cff')
process.load('Configuration.StandardSequences.EndOfProcess_cff')
process.load('Configuration.StandardSequences.FrontierConditions_GlobalTag_cff')

process.maxEvents = cms.untracked.PSet(
    input = cms.untracked.int32(100)
)

# Input source
process.source = cms.Source("PoolSource",
    fileNames = cms.untracked.vstring('root://eospublic.cern.ch//eos/opendata/cms/Run2016G/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/35017A26-8C9D-204D-92B6-3ABFBBD4ADF3.root'),
    secondaryFileNames = cms.untracked.vstring()
)

process.options = cms.untracked.PSet(

)

# Production Info
process.configurationMetadata = cms.untracked.PSet(
    annotation = cms.untracked.string('nano_data_2016_UL nevts:100'),
    name = cms.untracked.string('Applications'),
    version = cms.untracked.string('$Revision: 1.19 $')
)

# Output definition

process.NANOAODSIMoutput = cms.OutputModule("NanoAODOutputModule",
    compressionAlgorithm = cms.untracked.string('LZMA'),
    compressionLevel = cms.untracked.int32(9),
    dataset = cms.untracked.PSet(
        dataTier = cms.untracked.string('NANOAODSIM'),
        filterName = cms.untracked.string('')
    ),
    fileName = cms.untracked.string('file:nano_data2016.root'),
    outputCommands = process.NANOAODSIMEventContent.outputCommands
)

# Additional output definition

# Other statements
from Configuration.AlCa.GlobalTag import GlobalTag
#process.GlobalTag = GlobalTag(process.GlobalTag, '106X_dataRun2_v37', '')
process.GlobalTag.connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/106X_dataRun2_v37.db')
process.GlobalTag.globaltag = '106X_dataRun2_v37'

# Path and EndPath definitions
process.nanoAOD_step = cms.Path(process.nanoSequence)
process.endjob_step = cms.EndPath(process.endOfProcess)
process.NANOAODSIMoutput_step = cms.EndPath(process.NANOAODSIMoutput)

# Schedule definition
process.schedule = cms.Schedule(process.nanoAOD_step,process.endjob_step,process.NANOAODSIMoutput_step)
from PhysicsTools.PatAlgos.tools.helpers import associatePatAlgosToolsTask
associatePatAlgosToolsTask(process)

#Setup FWK for multithreaded
process.options.numberOfThreads=cms.untracked.uint32(4)
process.options.numberOfStreams=cms.untracked.uint32(0)
process.options.numberOfConcurrentLuminosityBlocks=cms.untracked.uint32(1)

# customisation of the process.


# End of customisation functions

# Customisation from command line

process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)))
# Add early deletion of temporary data products to reduce peak memory need
from Configuration.StandardSequences.earlyDeleteSettings_cff import customiseEarlyDelete
process = customiseEarlyDelete(process)
# End adding early deletion
```

Note that it will not fail if a frontier connection is available, only when not (to come to the original question)

**Comment by mbluj** [comment]
Ah, OK. How can I remove connection to frontier w/o loosing connection to GT?

**Comment by mbluj** [comment]
> Ah, OK. How can I remove connection to frontier w/o loosing connection to GT?

I see it in your cfg file. I will test it later, but now I must run.

**Comment by mbluj** [comment]
With configuration copied from above which have connection to conditions in sqlite file, giving python dump like this:
```
>>> process.GlobalTag
cms.ESSource("PoolDBESSource",
    DBParameters = cms.PSet(
        authenticationPath = cms.untracked.string(''),
        authenticationSystem = cms.untracked.int32(0),
        messageLevel = cms.untracked.int32(0),
        security = cms.untracked.string('')
    ),
    DumpStat = cms.untracked.bool(False),
    ReconnectEachRun = cms.untracked.bool(False),
    RefreshAlways = cms.untracked.bool(False),
    RefreshEachRun = cms.untracked.bool(False),
    RefreshOpenIOVs = cms.untracked.bool(False),
    connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/106X_dataRun2_v37.db'),
    globaltag = cms.string('106X_dataRun2_v37'),
    pfnPostfix = cms.untracked.string(''),
    pfnPrefix = cms.untracked.string(''),
    snapshotTime = cms.string(''),
    toGet = cms.VPSet()
)
``` 
it still works for me...

**Comment by mmusich** [comment]
for the sake of honesty,

> For the sake of brevity, removing the customization

this is not exactly removing the customization, you are still checking out a package that is not available in release via:
 
`git clone https://github.com/cms-opendata-analyses/PFNanoProducerTool.git PhysicsTools/PFNano` 

in a self-contained `CMSSW_10_6_30` (which is what is normally accepted as centrally supported) the configuration above doesn't run.

**Comment by katilp** [comment]
> > Ah, OK. How can I remove connection to frontier w/o loosing connection to GT?
> 
> I see it in your cfg file. I will test it later, but now I must run.

No, connecting to the /cvmfs area for condition data is not enough, is does not cut the frontier connection. That's why this goes unobserved and it only fails when there is no frontier connection at all. i.e CMS open data VM, or in the CMS open data docker container if one explicitly removes the frontier connection, see https://cms-opendata-releaseguide.docs.cern.ch/computing_environment/containers/#testing-without-frontier-connection

**Comment by katilp** [comment]
> for the sake of honesty,
> 
> > For the sake of brevity, removing the customization
> 
> this is not exactly removing the customization, you are still checking out a package that is not available in release via:
> 
> `git clone https://github.com/cms-opendata-analyses/PFNanoProducerTool.git PhysicsTools/PFNano`
> 
> in a self-contained `CMSSW_10_6_30` (which is what is normally accepted as centrally supported) the configuration above doesn't run.

Yes, of course, this is what we provide for open data users. We do not point them to anything centrally supported but provide them examples of how they can use CMS open data. We believe that this an issue independent from the 

> With configuration copied from above which has connection to conditions in a sqlite file, giving python dump like this:
> 
> ```
> >>> process.GlobalTag
> cms.ESSource("PoolDBESSource",
>     DBParameters = cms.PSet(
>         authenticationPath = cms.untracked.string(''),
>         authenticationSystem = cms.untracked.int32(0),
>         messageLevel = cms.untracked.int32(0),
>         security = cms.untracked.string('')
>     ),
>     DumpStat = cms.untracked.bool(False),
>     ReconnectEachRun = cms.untracked.bool(False),
>     RefreshAlways = cms.untracked.bool(False),
>     RefreshEachRun = cms.untracked.bool(False),
>     RefreshOpenIOVs = cms.untracked.bool(False),
>     connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/106X_dataRun2_v37.db'),
>     globaltag = cms.string('106X_dataRun2_v37'),
>     pfnPostfix = cms.untracked.string(''),
>     pfnPrefix = cms.untracked.string(''),
>     snapshotTime = cms.string(''),
>     toGet = cms.VPSet()
> )
> ```
> 
> it still works for me...

Yes, it will work if you do not cut the frontier connection. This is the issue. It goes unobserved.

**Comment by katilp** [comment]
> > for the sake of honesty,
> > > For the sake of brevity, removing the customization
> > 
> > 
> > this is not exactly removing the customization, you are still checking out a package that is not available in release via:
> > `git clone https://github.com/cms-opendata-analyses/PFNanoProducerTool.git PhysicsTools/PFNano`
> > in a self-contained `CMSSW_10_6_30` (which is what is normally accepted as centrally supported) the configuration above doesn't run.
> 
> Yes, of course, this is what we provide for open data users. We do not point them to anything centrally supported but provide them examples of how they can use CMS open data. We believe that this an issue independent from the
> 
> > With configuration copied from above which has connection to conditions in a sqlite file, giving python dump like this:
> > ```
> > >>> process.GlobalTag
> > cms.ESSource("PoolDBESSource",
> >     DBParameters = cms.PSet(
> >         authenticationPath = cms.untracked.string(''),
> >         authenticationSystem = cms.untracked.int32(0),
> >         messageLevel = cms.untracked.int32(0),
> >         security = cms.untracked.string('')
> >     ),
> >     DumpStat = cms.untracked.bool(False),
> >     ReconnectEachRun = cms.untracked.bool(False),
> >     RefreshAlways = cms.untracked.bool(False),
> >     RefreshEachRun = cms.untracked.bool(False),
> >     RefreshOpenIOVs = cms.untracked.bool(False),
> >     connect = cms.string('sqlite_file:/cvmfs/cms-opendata-conddb.cern.ch/106X_dataRun2_v37.db'),
> >     globaltag = cms.string('106X_dataRun2_v37'),
> >     pfnPostfix = cms.untracked.string(''),
> >     pfnPrefix = cms.untracked.string(''),
> >     snapshotTime = cms.string(''),
> >     toGet = cms.VPSet()
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it still works for me...
> 
> Yes, it will work if you do not cut the frontier connection. This is the issue. It goes unobserved.

If you cannot cut the frontier connection, take the RecoTauTag package, add prints to [this file](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_30/RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py#L5) locally, and you will see it ending up there. And that's all what I'm trying to say.

**Comment by mmusich** [comment]
> If you cannot cut the frontier connection, take the RecoTauTag package, add prints to [this file](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_30/RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py#L5) locally, and you will see it ending up there. And that's all what I'm trying to say.

I can indeed make the process crash by short-circuiting [this](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_30/RecoTauTag/Configuration/python/loadRecoTauTagMVAsFromPrepDB_cfi.py#L5) via this recipe:

```
cmsrel CMSSW_10_6_30
cd CMSSW_10_6_30/src
cmsenv
cmsDriver.py data_2016UL_OpenData --data --eventcontent NANOAODSIM --datatier NANOAODSIM --step NANO --conditions 106X_dataRun2_v37 --era Run2_2016,run2_nanoAOD_106Xv2 --customise_commands="process.add_(cms.Service('InitRootHandlers', EnableIMT = cms.untracked.bool(False)));delattr(process, 'loadRecoTauTagMVAsFromPrepDB')" --nThreads 4 -n 100 --filein /store/data/Run2016H/JetHT/MINIAOD/UL2016_MiniAODv2-v2/130000/676E37D2-044C-D346-92D9-A127A55FD279.root --fileout file:nano_data2016_nopf.root --no_exec
voms-proxy-init --voms cms
cmsRun data_2016UL_OpenData_NANO.py
```

which is independent from the `PFNano` customization.
The issue persists in `CMSSW_14_0_0_pre2` so this issue is indeed not fully solved even in recent (pre-)releases.

**Comment by hqucms** [comment]
I did some more investigations for `CMSSW_14_0_0_pre2` based on the recipe from @mmusich. It turns out after changing two things I can get it to work:
1. Removing [these tasks](https://github.com/cms-sw/cmssw/blob/d0954a1f327baf260e0c6e29c8b8adb214800359/PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py#L291-L295) from `patTauMVAIDsTask`. It seems that they are not used anyhow as the output NANO content remain unchanged after removing them. Maybe @mbluj can confirm if they are indeed unused?
1. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.

For the open NANO release, I suppose the easiest way is to gather a list of the tags needed but not in the GT, and just dump/add them into a sqlite file? I dumped a list of tags being loaded manually in the 106X workflow. Probably not all of them are strictly needed, but have all of them should make things work w/o connecting to Frontier.

[tags.txt](https://github.com/cms-sw/cmssw/files/14082660/tags.txt)

**Comment by vlimant** [comment]
> 2\. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.

what GT have you used instead of `auto:run2_data` ?

**Comment by mbluj** [comment]
> I did some more investigations for `CMSSW_14_0_0_pre2` based on the recipe from @mmusich. It turns out after changing two things I can get it to work:
> 
> 1. Removing [these tasks](https://github.com/cms-sw/cmssw/blob/d0954a1f327baf260e0c6e29c8b8adb214800359/PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py#L291-L295) from `patTauMVAIDsTask`. It seems that they are not used anyhow as the output NANO content remain unchanged after removing them. Maybe @mbluj can confirm if they are indeed unused?
> 2. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.
> 
> For the open NANO release, I suppose the easiest way is to gather a list of the tags needed but not in the GT, and just dump/add them into a sqlite file? I dumped a list of tags being loaded manually in the 106X workflow. Probably not all of them are strictly needed, but have all of them should make things work w/o connecting to Frontier.
> 
> [tags.txt](https://github.com/cms-sw/cmssw/files/14082660/tags.txt)

Thanks @hqucms!
I think you are right about the tasks to be removed. It is actually what I mentioned earlier here (or in other parallel thread) that whole content of `taus_updatedMVAIds_cff.py` should be reviewed (I suppose that this is not needed anymore). I plan to do it in next few days and prepare a PR to master (and backports if needed).
I also agree that adding missing payloads to GT (thanks for the list!) is the quickest fix for open-data workflows as it does not require new CMSSW release.

**Comment by katilp** [comment]
> > I did some more investigations for `CMSSW_14_0_0_pre2` based on the recipe from @mmusich. It turns out after changing two things I can get it to work:
> > 
> > 1. Removing [these tasks](https://github.com/cms-sw/cmssw/blob/d0954a1f327baf260e0c6e29c8b8adb214800359/PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py#L291-L295) from `patTauMVAIDsTask`. It seems that they are not used anyhow as the output NANO content remain unchanged after removing them. Maybe @mbluj can confirm if they are indeed unused?
> > 2. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.
> > 
> > For the open NANO release, I suppose the easiest way is to gather a list of the tags needed but not in the GT, and just dump/add them into a sqlite file? I dumped a list of tags being loaded manually in the 106X workflow. Probably not all of them are strictly needed, but have all of them should make things work w/o connecting to Frontier.
> > [tags.txt](https://github.com/cms-sw/cmssw/files/14082660/tags.txt)
> 
> Thanks @hqucms! I think you are right about the tasks to be removed. It is actually what I mentioned earlier here (or in other parallel thread) that whole content of `taus_updatedMVAIds_cff.py` should be reviewed (I suppose that this is not needed anymore). I plan to do it in next few days and prepare a PR to master (and backports if needed). I also agree that adding missing payloads to GT (thanks for the list!) is the quickest fix for open-data workflows as it does not require new CMSSW release.

Thanks! For open data, we would prefer the cleanest solution i.e. a new release with the fixes and a new GT if that can happen very shortly. It will be more work for us now for changes in already prepared material, but in the future, it will avoid patches and additional explanations in the CMS open data tutorials and guides. What would be the estimated timescale?

**Comment by hqucms** [comment]
> > 2. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.
> 
> what GT have you used instead of `auto:run2_data` ?

@vlimant `auto:run2_data` points to `133X_dataRun2_v2` in `CMSSW_14_0_0_pre2`. It seems that the tags missing in `106X_dataRun2_v37` are introduced in ~113X (e.g., https://cms-conddb.cern.ch/cmsDbBrowser/search/Prod/RecoTauTag_antiElectronMVA6v3_noeveto_gbr_NoEleMatch_woGwoGSF_BL).

**Comment by hqucms** [comment]
> > > I did some more investigations for `CMSSW_14_0_0_pre2` based on the recipe from @mmusich. It turns out after changing two things I can get it to work:
> > > 
> > > 1. Removing [these tasks](https://github.com/cms-sw/cmssw/blob/d0954a1f327baf260e0c6e29c8b8adb214800359/PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py#L291-L295) from `patTauMVAIDsTask`. It seems that they are not used anyhow as the output NANO content remain unchanged after removing them. Maybe @mbluj can confirm if they are indeed unused?
> > > 2. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.
> > > 
> > > For the open NANO release, I suppose the easiest way is to gather a list of the tags needed but not in the GT, and just dump/add them into a sqlite file? I dumped a list of tags being loaded manually in the 106X workflow. Probably not all of them are strictly needed, but have all of them should make things work w/o connecting to Frontier.
> > > [tags.txt](https://github.com/cms-sw/cmssw/files/14082660/tags.txt)
> > 
> > 
> > Thanks @hqucms! I think you are right about the tasks to be removed. It is actually what I mentioned earlier here (or in other parallel thread) that whole content of `taus_updatedMVAIds_cff.py` should be reviewed (I suppose that this is not needed anymore). I plan to do it in next few days and prepare a PR to master (and backports if needed). I also agree that adding missing payloads to GT (thanks for the list!) is the quickest fix for open-data workflows as it does not require new CMSSW release.
> 
> Thanks! For open data, we would prefer the cleanest solution i.e. a new release with the fixes and a new GT if that can happen very shortly. It will be more work for us now for changes in already prepared material, but in the future, it will avoid patches and additional explanations in the CMS open data tutorials and guides. What would be the estimated timescale?

I think if we create a dedicated sqlite or make a new GT to include the missing tags, then no change is needed for the release. And since the open data will be using a sqlite anyhow, it might be much easier to just add the missing tags to the sqlite, rather than making a new GT. Maybe the Alca/DB group can comment on this?

**Comment by katilp** [comment]
> > > > I did some more investigations for `CMSSW_14_0_0_pre2` based on the recipe from @mmusich. It turns out after changing two things I can get it to work:
> > > > 
> > > > 1. Removing [these tasks](https://github.com/cms-sw/cmssw/blob/d0954a1f327baf260e0c6e29c8b8adb214800359/PhysicsTools/NanoAOD/python/taus_updatedMVAIds_cff.py#L291-L295) from `patTauMVAIDsTask`. It seems that they are not used anyhow as the output NANO content remain unchanged after removing them. Maybe @mbluj can confirm if they are indeed unused?
> > > > 2. Switch the GT from `106X_dataRun2_v37` to a more recent one (I just used `auto:run2_data`). It seems that some of MVA tags for boosted taus are only included since 113X GTs.
> > > > 
> > > > For the open NANO release, I suppose the easiest way is to gather a list of the tags needed but not in the GT, and just dump/add them into a sqlite file? I dumped a list of tags being loaded manually in the 106X workflow. Probably not all of them are strictly needed, but have all of them should make things work w/o connecting to Frontier.
> > > > [tags.txt](https://github.com/cms-sw/cmssw/files/14082660/tags.txt)
> > > 
> > > 
> > > Thanks @hqucms! I think you are right about the tasks to be removed. It is actually what I mentioned earlier here (or in other parallel thread) that whole content of `taus_updatedMVAIds_cff.py` should be reviewed (I suppose that this is not needed anymore). I plan to do it in next few days and prepare a PR to master (and backports if needed). I also agree that adding missing payloads to GT (thanks for the list!) is the quickest fix for open-data workflows as it does not require new CMSSW release.
> > 
> > 
> > Thanks! For open data, we would prefer the cleanest solution i.e. a new release with the fixes and a new GT if that can happen very shortly. It will be more work for us now for changes in already prepared material, but in the future, it will avoid patches and additional explanations in the CMS open data tutorials and guides. What would be the estimated timescale?
> 
> I think if we create a dedicated sqlite or make a new GT to include the missing tags, then no change is needed for the release. And since the open data will be using a sqlite anyhow, it might be much easier to just add the missing tags to the sqlite, rather than making a new GT. Maybe the Alca/DB group can comment on this?

I think they commented in https://cms-talk.web.cern.ch/t/condition-database-access-outside-of-gt-for-nano-production/33715/5
In any case, a change is needed in the code either to remove the tasks or remove the frontier connection, and from the open data point of view we would prefer it clean. Open data will be using this release for years to come.

**Comment by mbluj** [comment]
Hello, sorry for delay in answering, but I was taken by other commitments. So, to summarize what needs to be done:
1.  Removal of tauIDs with payloads not in GT from official workflows:
it is quite straightforward in master, but requires more work in UL release series (10_6) where nothing in this direction has been performed as far as I remember. 
2. I expect that cleaning in the UL/10_6 releases will anyway require update of GT as I suppose that corresponding GT do not contain even minimal required set of needed payloads - to be checked.
3. What should be done with intermediate release series (>10_6 & <14_0)? Do we want backport of the cleaning to all of them? I would like to avoid it if not strictly necessary.

About timescale: I have other things on the plate (e.g. some L1 stuff for 2024 & phase-2), but I can reorder this if necessary. I suppose that cleaning master will take 1-2 days, backport to 10_6 a few additional days. What concerns GT update, I have not experience with it, so I have to either learn how to do or someone else should do the job - any help will be appreciated.

**Comment by perrotta** [comment]
Thank you @mbluj 

If I understand it correctly we should:
- Remove tauIDs with payloads not in GT from the master 
- Add the missing payloads to the GTs for UL/10_6  (AlCa can do it)

If you remove the tauIDs also from UL/10_6 then no update of the GTs is needed, but as it will probably require modifying in a non trivial way a closed release this should be probably better avoided.

For the intermediate releases I would check one by one: probably we could avoid acting on all them, but it depends on what's intended to do with them.

**Comment by mbluj** [comment]
> Thank you @mbluj
> 
> If I understand it correctly we should:
> 
> * Remove tauIDs with payloads not in GT from the master

Correct. As far as I understand the problem it touches only NanoAOD workflows.

> * Add the missing payloads to the GTs for UL/10_6  (AlCa can do it)
Yes, but the number of payloads to add is potentially bigger comparing to what is in GT for master and other release series. 
> 
> If you remove the tauIDs also from UL/10_6 then no update of the GTs is needed, but as it will probably require modifying in a non trivial way a closed release this should be probably better avoided.

It should be checked. In principle a "full cleaning" can require changes in AOD/RECO, miniAOD and NanoAOD sequences, but I suppose that the changes will not affect data content at any of datatiers. The only effect can be removal of era-dependent modifications for compatibility with old (pre-UL) samples. Anyway, if I understand it correctly the idea is to update only NanoAOD sequences as it is not expected to produce new UL-like samples other than NanoAOD for OpenData purposes, right? 


> 
> For the intermediate releases I would check one by one: probably we could avoid acting on all them, but it depends on what's intended to do with them.

OK. Changes in intermediate releases newer than 11_3 (if I am correct) will be similar to those in master while for older similar to those for 10_6. But, even trivial backporting and testing (sometimes creating GT) to a number of release series is already some additional burden.

**Comment by jmhogan** [comment]
Pinging this thread along with #43797 (@mbluj)

In the Open Data context, DPOA's preferred solution is cleaning these IDs from UL/10_6 so that a new release can be used without a new (bigger) GT.

Nano sequences are very likely to be the most used in Open Data, but we do provide instructions on how users can produce their own MC, which follows the full sequence. It would be ideal to clean this out fully. But the Nano sequences have the highest priority. We can test where in the full production chain it fails if needed.

**Comment by vlimant** [comment]
please close

done with #44685 until further notice
